Nos/Service/ReportPublisher.swift

```diff
@@ -0,0 +1,130 @@
+        switch target {
+        case .note(let note):
+            guard let keyPair = currentUser.keyPair else {
+                Log.error("Cannot publish report - No signed in user")
+                return
+            }
+            
+            guard let reportRequestDM = createReportRequestDM(note: note, category: category, keyPair: keyPair) else {
+                Log.error("Failed to create gift wrapped report request")
+                return
+            }
+            
+            Task {
+                do {
+                    try await relayService.publish(
+                        event: reportRequestDM,
+                        to: Relay.nosAddress,
+                        context: context
+                    )
+                    analytics.reported(target)
+                } catch {
+                    Log.error("Failed to publish report: \(error.localizedDescription)")
+                }
+            }
+        case .author:
+            Log.error("Private reports for people are no implemented yet")
+            return
+        }
+    }
+    
+    // MARK: Helpers
+    
+    internal struct ReportRequest {
+        let reportedEvent: JSONEvent
+        let reporterPubkey: RawAuthorID
+        let reporterText: String
+        
+        func toJSON() throws -> String {
+            let dictionary: [String: Any] = [
+                "reportedEvent": reportedEvent.dictionary,
+                "reporterPubkey": reporterPubkey,
+                "reporterText": reporterText
+            ]
+            
+            let data = try JSONSerialization.data(withJSONObject: dictionary)
+            return String(data: data, encoding: .utf8)!
+        }
+    }
+    
+    internal func createPublicReport(
+        for target: ReportTarget,
+        category: ReportCategory,
+        pubKey: RawAuthorID
+    ) -> JSONEvent {
+        var event = JSONEvent(
+            pubKey: pubKey,
+            kind: .report,
+            tags: [
+                ["L", "MOD"],
+                ["l", "MOD>\(category.code)", "MOD"]
+            ],
+            content: String(localized: .localizable.reportEventContent(category.displayName))
+        )
+        
+        let nip56Reason = category.nip56Code.rawValue
+        event.tags += target.tags(for: nip56Reason)
+        
+        return event
+    }
+    
+    internal func createReportRequestDM(note: Event, category: ReportCategory, keyPair: KeyPair) -> JSONEvent? {
+        guard let jsonEvent = note.codable else {
+            Log.error("Could not encode note to JSON")
+            return nil
+        }
+        
+        do {
+            let reportRequestJSON = try ReportRequest(
+                reportedEvent: jsonEvent,
+                reporterPubkey: keyPair.publicKeyHex,
+                reporterText: category.displayName
+            ).toJSON()
+            
+            let reportRequestGiftWrap = try DirectMessageWrapper.wrap(
```

Comment:
I was thinking about the function more as a generic wrapper for text, literally as a direct message for chat communication. This is mainly thinking from the perspective of future usage for some private messages in the app because the type 14 rumor used internally is really a direct message kind use for just [that](https://github.com/vitorpamplona/nips/blob/sealed-dms/17.md).

What do you think if I create a new wrapReportRequest that accepts the `ReportRequest` and leave this one as it is?

Reply:
Oh I see. I was picturing someday having direct message UI that would compose a "direct message" event and then call this function and pass it in. But I see your point that this function _already_ is composing that event internally. I think this code is fine the way it is now.